### PR TITLE
io/os: Implement IsTerminal trait on Stdio

### LINKED ISSUE #80937: Use `atty` crate in libtest instead of rolling our own
cc #80888
cc @jyn514

This issue has been assigned to @mattwilkinsonn via [this comment](https://github.com/rust-lang/rust/issues/80937#issuecomment-1079303047).

### PR DESCRIPTION
**Pull request by ghost**
Simple change to use the 'atty' crate in the cli.rs file instead of a custom function. Fixes #80937

However, I couldn't get this to build locally while running `./x.py build --stage 0` and I don't know why. It gives me this error (edited out unrelated parts):

> Compiling atty v0.2.14
> error[E0463]: can't find crate for core
> error: aborting due to previous error
> error: could not compile atty

I think this is all that needs to be done here but I'm fairly new to Rust and completely new to contributing to the actual compiler so I'm not sure. Would appreciate any help!

### CODE REVIEW

**Comment by jyn514** [comment]
```
     Checking atty v0.2.14
error[E0463]: can't find crate for `core`

error: aborting due to previous error
```
atty needs to be built after core, not before. You may want to see how this is done for other crates like `libc`.

**Comment by Zireael13** [comment]
> ```
>      Checking atty v0.2.14
> error[E0463]: can't find crate for `core`
> 
> error: aborting due to previous error
> ```
> 
> atty needs to be built after core, not before. You may want to see how this is done for other crates like `libc`.

Hey @jyn514! Thanks for the help. I've just made a PR to the atty repo: https://github.com/softprops/atty/pull/45 to add the 'rustc-dep-of-std' feature to 'atty', basically just using the same setup as 'cfg-if': https://github.com/alexcrichton/cfg-if/blob/main/Cargo.toml. Once this is added then it should compile just fine, seems to be working locally for me.

**Comment by Mark-Simulacrum** [comment]
Hm, so I actually am not sure we want to do this, or at least, I'm a bit worried about adding a dependency to libtest which seems to have significantly more complicated logic without a better understanding of that logic and the reasons it exists (probably good ones!); I would also prefer to keep std/test dependencies to ones controlled by rust-lang so that we don't need to fork things or otherwise have trouble editing them. It is also my recollection that there are a number of different crates which do this, in which case it seems reasonable to ask why this one was chosen :)

cc @rust-lang/libs - can you give your opinion here?

**Comment by m-ou-se** [comment]
I'd prefer to not add dependencies to the std/alloc/core/test/.. crates from outside the `rust-lang`. I wouldn't be opposed to adding `isatty()` to the standard library though.

**Comment by BurntSushi** [comment]
> Hm, so I actually am not sure we want to do this, or at least, I'm a bit worried about adding a dependency to libtest which seems to have significantly more complicated logic without a better understanding of that logic and the reasons it exists (probably good ones!);

I'm the one who wrote most of that more complicated logic and contributed it to the `atty` crate. I got it from how `git` does the same thing. The comments in the code of the `atty` crate explain most of it I think. The high level bit is that on Windows, detecting a tty differs depending on whether you're in a standard Windows console (such as `cmd.exe`) or whether you're in an MSYS terminal.

> I would also prefer to keep std/test dependencies to ones controlled by rust-lang so that we don't need to fork things or otherwise have trouble editing them.

I don't have any strong opinions on this, but would tend to agree with you. Certainly, I employ this strategy in some cases with my own crates, so it makes sense to me.

> It is also my recollection that there are a number of different crates which do this, in which case it seems reasonable to ask why this one was chosen :)

The `isatty` crate is deprecated in favor of `atty`. Maybe there are other tty detection crates that have `atty`'s logic now, but `atty` was definitely the first. Given the simplicity of the API, I've never done a survey of the ecosystem since `atty` did what I wanted. But my sense is that `atty` is the de facto standard solution to the problem in the ecosystem today.

-----

Other thoughts:

1. It might be okay to just copy the logic from `atty` instead of bringing in the crate itself. The logic has been rather fixed and unchanged for a few years at this point. While some of the details are complex, the amount of code is very small.
2. I would also be in favor of adding an `isatty` routine to std as well, although, the way in which the `atty` crate detects a tty on Windows is pretty non-standard. In particular, its cygwin/MSYS handling is mostly a heuristic. So whether that would be appropriate for a std API isn't quite clear to me. It depends on whether we can get away with documenting it as a platform specific heuristic that can't be relied on 100%.

**Comment by Amanieu** [comment]
I feel that this logic should probably be moved into libstd since we want `Stdout` to only be line buffered if we are printing to a tty.

https://github.com/rust-lang/rust/blob/a118ee2c13cc96ceb27bd5030c1cca1052377604/library/std/src/io/stdio.rs#L491-L494

**Comment by mattwilkinsonn** [comment]
Coming back to this after getting some more Rust experience.

So basically we'd want to just take the logic that is currently in the `atty` crate and put it into `std/io`? If so, should that go in the `stdio.rs` file, in a new file, or somewhere else?

Would appreciate any help on this!

**Comment by Amanieu** [comment]
I feel that the best place for the logic is in `library/std/src/sys/$TARGET/stdio.rs`.

This then needs to be exported as an unstable feature in `std`'s public API so that `test` can import it. Perhaps this could be a method on the `Stdin`/`Stdout`/`Stderr` types?

**Comment by mattwilkinsonn** [comment]
> I feel that the best place for the logic is in `library/std/src/sys/$TARGET/stdio.rs`.
> 
> This then needs to be exported as an unstable feature in `std`'s public API so that `test` can import it. Perhaps this could be a method on the `Stdin`/`Stdout`/`Stderr` types?

Hey @Amanieu , Had a first go at implementing the methods on all of the platforms that the `atty` crate supported (unix, windows, hermit) . I added an `IsAtty` trait to `std::io` that I just repeated on each of the three streams for each of the platforms.

I marked it all with an unstable feature `is_atty` and also had to mark everything that wasn't part of the feature as stable so the rest of the project still complied. Not sure if that was correct though, especially because I just put it as:
 `#[stable(feature = "rust1", since = "1.0.0")] `
because I don't actually know when these were stabilized. I'm guessing it was pre-1.0 though.

Let me know what you think, I'm sure I did something wrong, lol.

File: library/std/src/lib.rs [hunk 1]
```diff
@@ -516,6 +516,9 @@ pub mod task {
 mod sys_common;
 mod sys;
 
+#[unstable(feature = "is_atty", issue = "80937")]
+pub use sys::stdio::{Stderr, Stdin, Stdout};
```
Comment (bjorn3) [inline library/std/src/lib.rs:520]: None of the items inside `sys` should be exported. This is solely an implementation detail. Instead `is_atty` should be on the `Std*` types inside `std::io`. You could add an `IsAtty` trait to each of the relevant platform modules inside the `os` module and implement it there. This should likely forward to a method on `sys::stdio::Std*` though. The `os` module is where platform specific extensions should be.

File: library/test/src/lib.rs [hunk 2]
```diff
@@ -290,7 +291,11 @@ where
     fn calc_timeout(timeout_queue: &VecDeque<TimeoutEntry>) -> Option<Duration> {
         timeout_queue.front().map(|&TimeoutEntry { timeout: next_timeout, .. }| {
             let now = Instant::now();
-            if next_timeout >= now { next_timeout - now } else { Duration::new(0, 0) }
+            if next_timeout >= now {
+                next_timeout - now
+            } else {
+                Duration::new(0, 0)
+            }
```
Comment (bjorn3) [inline library/test/src/lib.rs:298]: I think you have been formatting with a stable rustfmt that doesn't support some of the options specified in `rustfmt.toml`. Please use `./x.py fmt` instead.

File: library/test/src/cli.rs [hunk 3]
```diff
@@ -30,7 +31,7 @@ pub struct TestOpts {
 impl TestOpts {
     pub fn use_color(&self) -> bool {
         match self.color {
-            ColorConfig::AutoColor => !self.nocapture && isatty::stdout_isatty(),
+            ColorConfig::AutoColor => !self.nocapture && Stdout::is_atty(),
```
Comment (bjorn3) [inline library/test/src/cli.rs:34]: As written this will break all platforms for which there is no `IsAtty` implementation yet.

**Comment by jyn514** [comment]
@mattwilkinsonn did you mean to close this?

**Comment by mattwilkinsonn** [comment]
> @mattwilkinsonn did you mean to close this?

No was reverting my commits so I can get back on the latest version of the repo. Will reopen once i have another revision

**Comment by mattwilkinsonn** [comment]
@bjorn3 
Looking in `std::os`, seems like the std::os::windows crate is actually std::sys::windows::ext, just re-exported. I assume I should implement the trait in there? Same for unix & hermit.

**Comment by bjorn3** [comment]
Ah, yes. I looked at the linux specific mod which isn't a re-export of `...::ext`.

**Comment by mattwilkinsonn** [comment]
@bjorn3

Ok, I implemented the trait inside the `ext` modules for each platform on the raw `Stdout`, `Stdin`, and `Stderr`, and then wrapped around those in `io`, using a cfg_if to prevent issues for the other platforms. Let me know what you think

File: library/std/src/io/stdio.rs [hunk 4]
```diff
@@ -965,3 +965,57 @@ pub fn _eprint(args: fmt::Arguments<'_>) {
 
 #[cfg(test)]
 pub use realstd::io::{_eprint, _print};
+
+/// Trait to determine if stdio stream (Stdin, Stdout, Stderr) is a tty.
+#[unstable(feature = "is_atty", issue = "80937")]
+pub trait IsAtty {
```
Comment (Amanieu) [inline library/std/src/io/stdio.rs:971]: Slight nit: this is actually 3 words: is a TTY

So the proper name for the trait should be `IsATty`. But that looks pretty ugly so perhaps `IsTerminal` might be a better name.

Comment (mattwilkinsonn) [reply library/std/src/io/stdio.rs]: Renamed to IsTerminal and renamed the function and feature accordingly

**Comment by m-ou-se** [comment]
I see this is still marked as draft. What work is there left to do before this is ready for review?

**Comment by bjorn3** [comment]
There are compile error it seems.

File: library/std/src/sys/windows/ext/io.rs [hunk 5]
```diff
@@ -220,3 +220,141 @@ impl IntoRawSocket for net::UdpSocket {
...
+            STD_ERROR_HANDLE as STD_ERROR, STD_INPUT_HANDLE as STD_INPUT,
+            STD_OUTPUT_HANDLE as STD_OUTPUT,
+        };
+
+        let fd = STD_OUTPUT;
+        let others = [STD_INPUT, STD_ERROR];
+
+        if unsafe { console_on_any(&[fd]) } {
+            // False positives aren't possible. If we got a console then
+            // we definitely have a tty on stdin.
```
Comment (r00ster91) [inline library/std/src/sys/windows/ext/io.rs:268]:
```suggestion
            // we definitely have a tty on stdout.
```

File: library/std/src/sys/windows/ext/io.rs [hunk 6]
```diff
@@ -220,3 +220,141 @@ impl IntoRawSocket for net::UdpSocket {
...
+            STD_ERROR_HANDLE as STD_ERROR, STD_INPUT_HANDLE as STD_INPUT,
+            STD_OUTPUT_HANDLE as STD_OUTPUT,
+        };
+
+        let fd = STD_ERROR;
+        let others = [STD_INPUT, STD_OUTPUT];
+
+        if unsafe { console_on_any(&[fd]) } {
+            // False positives aren't possible. If we got a console then
+            // we definitely have a tty on stdin.
```
Comment (r00ster91) [inline library/std/src/sys/windows/ext/io.rs:299]:
```suggestion
            // we definitely have a tty on stderr.
```

File: library/std/src/sys/windows/ext/io.rs [hunk 7]
```diff
@@ -220,3 +220,141 @@ impl IntoRawSocket for net::UdpSocket {
...
     }
 }
+
+#[unstable(feature = "is_terminal", issue = "80937")]
+impl io::IsTerminal for sys::stdio::Stdin {
+    fn is_terminal() -> bool {
+        use c::{
+            STD_ERROR_HANDLE as STD_ERROR, STD_INPUT_HANDLE as STD_INPUT,
+            STD_OUTPUT_HANDLE as STD_OUTPUT,
+        };
```
Comment (r00ster91) [inline library/std/src/sys/windows/ext/io.rs:230]: Is it a possibility to just remove these 3 `use`s and use them directly like `c::STD_ERROR_HANDLE` etc. like it's done in other parts of this file? See e.g. https://github.com/rust-lang/rust/blob/9213d20487b9f5d3fb538cb16b95fd8118b024dc/library/std/src/sys/windows/ext/io.rs#L70

**Review by ghost — CHANGES_REQUESTED** [review]
I would really like to see this move forward. I think it'd be great to have this in the `std`.

**Comment by mattwilkinsonn** [comment]
Sorry, got distracted with other stuff for the past few months. I'll take a look at this again next week. Thanks for the feedback!

**Comment by mattwilkinsonn** [comment]
I've been looking into making these updates, however it looks like my branch is far enough behind that I can't get the submodule artifacts in order to run any `x.py` commands. 
image

In addition rebasing onto thousands of commits sounds a bit daunting. What would be the best approach for this? Just revert back to the HEAD and manually rewrite the changes into a new commit?

**Comment by Amanieu** [comment]
Considering the changes aren't too big (only ~250 lines changed) it's reasonable to attempt a manual rewrite if you can't get rebasing to work.

**Comment by mattwilkinsonn** [comment]
Closing as I've reimplemented the changes in #91121.

**Pull request closed without merging**